lib/stages/integrateHFBData.cpp

```diff
@@ -0,0 +1,228 @@
+    total_lost_timesamples += get_fpga_seq_num(in_buf, in_buffer_ID) - fpga_seq_num_start;
+    // Get the first FPGA sequence no. to check for missing frames
+    fpga_seq_num = get_fpga_seq_num(in_buf, in_buffer_ID);
+    frame++;
+
+    DEBUG("\nInit frame. fpga_seq_num: {:d}. sum_data[0]: {:f}",
+          get_fpga_seq_num(in_buf, in_buffer_ID), sum_data[0]);
+}
+
+void integrateHFBData::integrateFrame(float* input_data, float* sum_data,
+                                      const uint32_t in_buffer_ID) {
+    frame++;
+    fpga_seq_num += _samples_per_data_set;
+    total_lost_timesamples += get_fpga_seq_num(in_buf, in_buffer_ID) - fpga_seq_num;
+    fpga_seq_num = get_fpga_seq_num(in_buf, in_buffer_ID);
+
+    // Integrates data from the input buffer to the output buffer.
+    for (uint beam = 0; beam < _num_frb_total_beams; beam++) {
+        for (uint freq = 0; freq < _num_sub_freqs; freq++) {
+            sum_data[beam * _num_sub_freqs + freq] += input_data[beam * _num_sub_freqs + freq];
+        }
+    }
+
+    DEBUG("\nIntegrate frame {:d}, total_lost_timesamples: {:d}, sum_data[0]: {:f}\n", frame,
+          total_lost_timesamples, sum_data[0]);
+}
+
+float integrateHFBData::normaliseFrame(float* sum_data, const uint32_t in_buffer_ID) {
+
+    const float normalise_frac =
+        (float)total_timesamples / (total_timesamples - total_lost_timesamples);
+
+    for (uint beam = 0; beam < _num_frb_total_beams; beam++) {
+        for (uint freq = 0; freq < _num_sub_freqs; freq++) {
+            sum_data[beam * _num_sub_freqs + freq] *= normalise_frac;
+        }
+    }
+
+    DEBUG("Integration completed with {:d} lost samples", total_lost_timesamples);
+
+    fpga_seq_num = get_fpga_seq_num(in_buf, in_buffer_ID);
+
+    return normalise_frac;
+}
+
+void integrateHFBData::main_thread() {
+
+    uint in_buffer_ID = 0,
+         compress_buffer_ID = 0; // Process only 1 GPU buffer, cycle through buffer depth
+    uint8_t* in_frame;
+    uint8_t* compressed_lost_samples_frame;
+    int out_buffer_ID = 0, first = 1;
+    int64_t fpga_seq_num_end_old;
+    total_timesamples = _samples_per_data_set * _num_frames_to_integrate;
+    total_lost_timesamples = 0;
+    fpga_seq_num = 0, fpga_seq_num_end = (_num_frames_to_integrate - 1) * _samples_per_data_set;
+    frame = 0;
+
+    // Get the first output buffer which will always be id = 0 to start.
+    uint8_t* out_frame = wait_for_empty_frame(out_buf, unique_name.c_str(), out_buffer_ID);
+    if (out_frame == NULL)
+        goto end_loop;
+
+    while (!stop_thread) {
+        // Get an input buffer, This call is blocking!
+        in_frame = wait_for_full_frame(in_buf, unique_name.c_str(), in_buffer_ID);
+        if (in_frame == NULL)
+            goto end_loop;
+
+        compressed_lost_samples_frame = wait_for_full_frame(
+            compressed_lost_samples_buf, unique_name.c_str(), compress_buffer_ID);
+        if (compressed_lost_samples_frame == NULL)
+            goto end_loop;
```

Comment:
Just to make sure we write it down, currently this could result in a desync between the `in_buf` and `compressed_lost_samples_buf` if some frames go missing in one buffer but not the other.   You might want to see what Richard did for this in `RfiFrameDrop`, it doesn't involve lost samples, but it does cover this case.   I think Davor also did something like this in one of the frb/pulsar post process stages.   Given this pattern has shown up a few times, we might want to abstract this pattern into its own stage.